### DISCUSSION TITLE
Translation of the user guide

### DISCUSSION THREAD

**Discussion by Cedric-1515** [vkbo/novelWriter#1834] [Ideas]
Hi,
I'm a new user of novelWrite and already enthusiastically using it. Talked about it to a couple of people who are not speaking English and they remain kind of reluctant to skip from their usual commercial docx software to an open source text editor.
Is there anyone willing or planning to translate the user guide? 
I can contribute to a French version.

**Comment by vkbo** [comment]
Someone contributed a French translation of the docs a couple of years ago, but unfortunately keeping up with it is a huge task as it changes a fair bit between each release.

Ideally, a translated documentation would be nice though. The translation tool novelWriter uses is limited by word count, and there is no way the documentation is anywhere near small enough, so I'm uncertain what would be the best approach. Publishing alternative PDF docs could be one option, but without a tool that checks what paragraphs need re-translating for each version, the task is pretty big. If you know of a suitable tool, I'm happy to look into it.

I could set up a shadow repository where translated version of the documentation lives, and automatically build PDFs, so that part is definitely doable.

**Reply by Cedric-1515** [reply]
I'm sorry I have no idea which tools would be best for the task. Of course, the more people involved, the better and faster it can be finalized. 

If there is no specific solution for the moment nor anyone else ready to collaborate, I don't mind translating it myself. 

Hope I guess right, you edited the documentation with LaTeX?  I can work from the original files and send them back to you once done. I'm currently working on multilingual documents with LaTeX. The advantage would be that I don't need to change anything about the coding parts nor the images, just translating text, titles and captions and compiling to build the PDF.

**Reply by vkbo** [reply]
The documentation is written with restructuredtext (.rst files) which can generate multiple different outputs. I currently build HTML docs for the website, and LaTeX for the PDFs.

The source for the docs currently lives in [the docs folder of the repo](https://github.com/vkbo/novelWriter/tree/main/docs/source). It is copied to the main website on release, and the PDF manual is also built automatically and included in release packages.

The documentation is written for Sphinx, and there is a translation procedure ([see their docs](https://www.sphinx-doc.org/en/master/usage/advanced/intl.html)), and I'm happy to set it up. As I said, I would probably want to do this in a separate repository. Using a separate repository will ensure that the translation work is only needed once I publish a new version, and not every time there is a small change made.

**Reply by vkbo** [reply]
I would not recommend translating the raw text directly. That makes it very difficult to update when there are changes. The translation tool instead does this one paragraph at a time.

**Reply by Cedric-1515** [reply]
Oh okay.. Well, I got to go into the documentation of Sphinx as I never used it before and keep you updated

**Reply by Cedric-1515** [reply]
Here I am again. I created an account on transifex and checked how it works on the Sphinx translation project itself.

If I get it right, you set up a project there and instead of translating the raw files, I check one chain after the other on the platform and then I'm ready to go?

**Reply by vkbo** [reply]
Never used the Transifex service, so I don't know. Usually even open source projects have a word count limit, and I expect they have one too. Usually it's around the lowest tier of paid plans. Here that is about 50000 words, which gets us exactly nowhere. The current documentation is around 27000 words.

The main application uses Crowdin for translations, and  this service has a 60000 word limit, of which we're currently using about 40000 for 13 languages.

**Reply by Cedric-1515** [reply]
According to the Transifex website (https://www.transifex.com/open-source/), they provide their service for free to OpenSource projects. "Localize your Open Source projects for free and reach new markets." It might be a solution. 

In the end, I'll agree on whichever application is fine for you.

I'm ready do the translation, which would be useful to the French-speaking community interested in a free software for their writing a novel.

**Reply by vkbo** [reply]
Yes, it's free for open source, But the question is what the word count limit is. Assuming they have one. I've looked all over their website and can't find any information either way.

If the word limit is anything like equivalent services, it won't be enough for even a single translation into French. This is precisely why I never set up a translation service for the documentation in the first place. But I can try to do some more research on this service later.

**Reply by vkbo** [reply]
I may add that finding information on that website is incredibly frustrating, and the website design is not at all reader-friendly with too many boxes flying around. I've ran out of patience with it for now. If you can find any information, please let me know.

**Reply by Cedric-1515** [reply]
Please keep me posted about it.
In the meantime, I'll start translating on a simple file as I have some free time I can dedicate to it right now and we see later how to find a solution. 
Looking forward to contributing to novelWriter's documentation!

**Reply by vkbo** [reply]
Well, translating files directly will not be very usable. The framework for translations uses files like this as the source files, which are the ones that need to be translated:

![image](https://github.com/vkbo/novelWriter/assets/1619840/f106e57b-16ac-4305-afa9-67110d9e8460)

**Reply by vkbo** [reply]
It's easy to generate the files, which I've already done. I've attached them here: [gettext.zip](https://github.com/vkbo/novelWriter/files/15133357/gettext.zip)

They are probably easier to translate with a suitable tool if you can find one. I have never translated this type of files before, so not very familiar with them.

**Reply by Cedric-1515** [reply]
I will make some research during the weekend. 

One last question, if I don't find any tool for it, would it make it useful for you if  I edit directly the .pot files with notepad, and then sending back to you a zipped folder with the translated files?

**Reply by vkbo** [reply]
You need to generate the .po files for the relevant language. See point 4 [here](https://www.sphinx-doc.org/en/master/usage/advanced/intl.html#quick-guide). I can easily do that for you as well, but it looks very clunky having to do all of this in notepad. There is a tool for it called [poedit](https://poedit.net/) available that does all that directly from the .pot files.

**Comment by Karduin** [comment]
Hi, I made a doc in French using sphinx, sphinx-intl, gettext, poedit and latex. You can find it here.

https://github.com/Karduin/novelWriter-docs_fr

In the ```/docs/fr_build/htm/``` folder: the web doc.

You can download the folder and run index.html.

In the ```/docs/fr_build/latex/``` folder: the pdf doc.

https://github.com/Karduin/novelWriter-docs_fr/blob/main/docs/fr_build/latex/manual.pdf

The .po files are in ```/docs/sources/locales/fr/LC_MESSAGES/```

I'm not sure if the translation is optimal, but it does the job.

**Reply by vkbo** [reply]
Thanks. I can certainly upload the PDF to the main website when I make the 2.6 release.

Do you have a suggestion for what would be a good way to maintain these translations in a more formal way? I suppose they can be kept in the main repo and updates to the translations can be accepted via pull requests.

**Reply by vkbo** [reply]
I looked at your repo, and as far as I can tell, all that is needed is to keep the `docs/source/locales/fr/LC_MESSAGES` folder and the `*.po` files within it. Everything else is auto-generated. Those could be added to the main documentation folder. As long as someone is willing to maintain the translations, I'm happy to publish them. But this is not a one time effort. The documentation changes on every release, although usually there aren't a lot of changes,

Edit: and the `_build/gettext/*.pot`  files are needed of course.

**Reply by Karduin** [reply]
> But this is not a one time effort.

you're right.

> all that is needed is to keep the docs/source/locales/fr/LC_MESSAGES folder and the *.po and _build/gettext/*.pot

I agree.

So i wonder if:

* Does `sphinx-intl update -p fr_build/gettext -l fr` update .po files without deleting the existing ones ?
* Who updates .po files before translation ? (you, i guess)
* I need to learn how to properly make a pull requests, any suggestions for further reading ?

**Reply by vkbo** [reply]
I've never done this either, but my understanding is that the *.pot files need to be updated when the documentation changes. I guess it's up to the translators to generate new *.po files and update the translations. I expect it will only update changes, not delete the old translations.

The HTML and PDF files are not needed. They can be generated on the fly when needed. Like during the release process.

**Reply by Karduin** [reply]
Sorry one more question,

Should I pull requests both folders and subfloders to the release branch or wait for you to do it?

**Reply by vkbo** [reply]
Let me set up the framework first, and then you can submit the translated files as a pull request. Not sure when I'll do it. I'm taking it slow over the holidays, but let's try to include it for the 2.6 RC 1 release.

**Reply by peter88213** [reply]
> my understanding is that the *.pot files need to be updated when the documentation changes. I guess it's up to the translators to generate new *.po files and update the translations. I expect it will only update changes, not delete the old translations.

I don't know how far the sphinx tools go, but the transition from the automatically generated template (.pot) to the actual dictionary (.po) is usually manual. I wrote myself a script for this which reads in both the updated .pot file and the last .po file and creates a new po file from it via a Python dictionary. In this .po file only the new and changed entries have to be filled in manually. 

It's quick to do, of course, but just in case, here's the link to my script:
https://github.com/peter88213/novelibre/blob/main/tools/translations.py

By the way, I collect all translations in a JSON database to be able to reuse existing translations across repositories, as I manage an application with a number of plugins where the terms should be used consistently.

**Reply by Karduin** [reply]
Thank's, It seems that sphinx and sphinx-intl do the job. I've modified an .rst file, (changed one line and added another).

Then
* run make gettext
* run sphinx-intl update -p _build/gettext -l fr
Update: source/locales/en/LC_MESSAGES/index.po +1, -1

I keep the unmodified translations and only edit the modified and new lines with poedit.

**Reply by peter88213** [reply]
> It seems that sphinx and sphinx-intl do the job.

Good to know. I'm interested in the topic because it could catch up with me at some point if others help with the translations in my projects. 

If this is of interest to you, here is my personal experience:
I use GNU gettext to translate my user interface, but not for the translation of my online manual, which I also have built with Sphinx. 

I wrote an English version first, which I later translated back into German. However, I decided against the gettext method supported by Sphinx, because then I would have to touch up the German translation for every minor correction to my English text, such as typos or poor wording (English is not my native language). 
For this reason, I maintain the two language versions of my manual separately, which mainly works because I can determine the further development of my software on my own. 

Only at the very beginning I used the “.po” files of my software to translate the fixed terms of the user interface with a search/replace script. Then I manually replaced the remaining English continuous text with my German translation. 

By the way, Microsoft doesn't seem to have done this with the last version of Word I used (95 or 97? or 2007?), but commissioned the translation of the help texts separately. In any case, not all menu items and technical terms in the help text corresponded to those in the user interface, which could be quite frustrating.

**Reply by vkbo** [reply]
A challenge with the documentation is definitely to keep the text in it in sync with the same labels on the GUI. This is manually taken care of in the default English GUI and docs. But for French, like here for instance, these could easily diverge when there are different translators for the GUI and the docs, and both translate from the respective English sources.

**Reply by Karduin** [reply]
Yes, so I worked with novelWriter open on the side. There are certainly a few things to correct, let's just say that as an amateur writer it's first draft. ;-)

Maybe I'll make a kind of dictionary with nw_fr_FR.ts

**Reply by peter88213** [reply]
This is a highly interesting topic, especially for software that is being developed rapidly but requires reliable documentation in order to use the features. Especially with free software, localizations are often hopelessly outdated. 

I have decided to only distribute the translations and docs I have created myself. My package builder aborts if the German translations are not complete. 

In the event that another translation is required, a maintainer is needed to take responsibility for it and the documentation. He can then offer a language package under his account that can be easily installed and is also upwardly compatible by simply leaving missing terms untranslated. I have marked the URL for the “Online help” menu entry for gettext so that it can refer directly to the translated help page in the context of a foreign language UI translation. 

Then I created a template repository for language packages on GitHub, which contains all the utilities to collect the _*.pot_ files, generate the necessary binaries from the _*.po_ files, assemble the whole thing into a self-installing package and link it to the localisation's landing page for download. The translated documentation should then be offered by this maintainer via a separate repository. 

At the moment I don't have a user base, especially as the use case for my program is quite specific, but I hope to be prepared for all possibilities in this way.

**Reply by peter88213** [reply]
> Maybe I'll make a kind of dictionary with nw_fr_FR.ts

I have got a Python script for this. Its output looks like so:
```markdown

# GuiDocEditor

Add Word to Dictionary	:	Ajouter ce mot au dictionnaire
Copy	:	Copier
Could not save document.	:	Impossible d'enregistrer le document.
Create Note for Tag	:	Créer une note pour l'étiquette
Created: {0}	:	Créé : {0}
Cut	:	Couper
Do you want to create a new project note for the tag '{0}'?	:	Voulez-vous créer une nouvelle note de projet pour l'étiquette '{0}' ?
Document Details	:	Détails du document
File Location: {0}	:	Emplacement du fichier : {0}
Follow Tag	:	Suivre cette étiquette
No Suggestions	:	Pas de suggestion

```

 Just uploaded it to the gists:
https://gist.github.com/peter88213/ce16d7df7c3351932af5776f4cd959f8

In case you need the translations as a JSON file:
https://gist.github.com/peter88213/7d0ca8f05fb1fa95eb02b39d356f0385

**Reply by Karduin** [reply]
Hi @peter88213
thank you for discovering the xml.etree module, which I didn't know. I made a SQLite database with a small Flask app, so I can search for terms in the source.

Well, maybe that's a bit too much, but I had fun.

![screen_nw](https://github.com/user-attachments/assets/d35a7a26-02ff-450f-92c0-77673e64d83e)

**Reply by peter88213** [reply]
This looks really nice. Is it also available to other translators?

**Reply by vkbo** [reply]
This looks great! Maybe you can assemble some documentation on this?

The 2.6 release is coming up very soon, so I think we can aim for the documentation update after that. That's mainly about including the PDF in the release. That'll have to wait until 2.7. But the website is independent of this, as it is in a separate repo, so adding the features needed for multi-language docs can be done independently.

**Reply by Karduin** [reply]
> This looks really nice. Is it also available to other translators?

Yes, all nw_xx_XX.ts language files in the i18n folder can be used. 

> This looks great! Maybe you can assemble some documentation on this?

everything needed is here, except the nw_xx_XX.ts files. There's also a short how-to.
https://github.com/Karduin/nw_gui_dict

@vkbo no worries, I'm patient

**Reply by vkbo** [reply]
If you generate the PDF for 2.6, I can manually include it this time, and we can aim to get it built automatically from 2.7? Since you've already done a lot of work that can benefit users.

I still haven't made all the documentation updates for 2.6, so there will be some changes probably. But likely not many. I'll keep you posted. I'll try to update the docs this weekend.

**Reply by Karduin** [reply]
all right, noted.

Furthermore, when I update local files, sphinx-intl indicates which files have been modified.

```
Not Changed: source/locales/fr/LC_MESSAGES/int_glossary.po
Not Changed: source/locales/fr/LC_MESSAGES/int_overview.po
Not Changed: source/locales/fr/LC_MESSAGES/int_howto.po
Not Changed: source/locales/fr/LC_MESSAGES/int_introduction.po
Not Changed: source/locales/fr/LC_MESSAGES/int_started.po
Not Changed: source/locales/fr/LC_MESSAGES/more_counting.po
Not Changed: source/locales/fr/LC_MESSAGES/more_projectformat.po
Not Changed: source/locales/fr/LC_MESSAGES/tech_locations.po
Not Changed: source/locales/fr/LC_MESSAGES/project_structure.po
Not Changed: source/locales/fr/LC_MESSAGES/project_manuscript.po
Not Changed: source/locales/fr/LC_MESSAGES/project_references.po
Not Changed: source/locales/fr/LC_MESSAGES/tech_tests.po
Not Changed: source/locales/fr/LC_MESSAGES/tech_storage.po
Not Changed: source/locales/fr/LC_MESSAGES/usage_breakdown.po
Not Changed: source/locales/fr/LC_MESSAGES/usage_project.po
Not Changed: source/locales/fr/LC_MESSAGES/usage_typography.po
Update: source/locales/fr/LC_MESSAGES/index.po +1, -1
Not Changed: source/locales/fr/LC_MESSAGES/usage_format.po
Not Changed: source/locales/fr/LC_MESSAGES/usage_shortcuts.po
Update: source/locales/fr/LC_MESSAGES/tech_source.po +2, -2
Update: source/locales/fr/LC_MESSAGES/more_customise.po +1, -1
Update: source/locales/fr/LC_MESSAGES/usage_writing.po +1, -2
Update: source/locales/fr/LC_MESSAGES/project_overview.po +4, -0
```

**Reply by vkbo** [reply]
Keep in mind that the `main` branch is now on release 2.7 Alpha. For the upcoming 2.6 release, it now lives in the `release` branch, I usually move it there when I'm done with the Beta and RC releases.

So the documentation updates for 2.6 will be on the `release` branch.

**Reply by Karduin** [reply]
Thank you, I will correct my mistake.

**Reply by Karduin** [reply]
Hi, if you want to use it, the French version 2.6 of the manual has been updated with the latest modifications here:

https://github.com/Karduin/novelWriter-docs_fr/blob/main/fr_manual_26.pdf

**Reply by vkbo** [reply]
Excellent. I was going to ask. I'll start looking into how to include it. I haven't yet decided what is the best approach. I think adding it to the install package and loading it if the GUI language is French is the best approach.

In any case, we can start looking at how to get your translation files into the main repository. You should be able to make your updates as pull requests to the main repo instead of using a separate one. I also need to figure out how to replace the "User Guide" title with a French title.

**Reply by vkbo** [reply]
Oh, and I will do a 2.6.1 release as soon as the other GUI translations are done, and I can include the manual then too.

**Reply by Karduin** [reply]
As you said earlier, since you generate everything automatically, you just need to keep the
* docs/source/locales/fr/LC_MESSAGES folder and the .po and .mo files
*  and _build/gettext/.pot.


I think you also need :

* Modify the config.py by adding:
```python
locale_dirs = ['locales/']
gettext_compact = False
```
* The make latex command replaces the contents of the latex folder. Is there a trick to keeping the original version and the translated version other than having a different build folder?

The .pot files must be common to all, so it seems logical to me that you should be the one to generate them. (But maybe I'm wrong) 

I think I should only PR  .po and .mo files.


What do you think?

**Reply by Karduin** [reply]
> I also need to figure out how to replace the "User Guide" title with a French title.

modify this line in config.py works.
```python
latex_documents = [(
    master_doc, "manual.tex", "Guide Utilisateur", author, "manual"
)]
```
I can recompile the pdf. It must be possible to condition compilation according to language.

**Reply by vkbo** [reply]
Yes, that I know. I just need to figure out how to make it switch based on language.

I can simply pull the name from the index.rst document, like explained [here](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-latex_documents) instead of hardcoding it. I'll look into that.

**Reply by vkbo** [reply]
I've now added the PDF directly to the repo as a preliminary solution for 2.6. I will add the translation framework in the main branch for the 2.7 release, and then remove the PDF from the repo so it can be built automatically when releases are made.

When the framework is set up in main, you should be able to make a pull request with the translation files to the `docs/source/locales/fr` folder. That way you get full credit for the contribution in the repo, and there is an official copy of the translation files that anyone can extend. I will add the needed commands to the pkgUtils script to handle the translation source generation and building. I need some custom code to copy the generated PDFs to the correct place too.

I'll tag you in the pull request where I set it up so you can test it and let me know if it works. I'll also look at how to integrate this into the official website. At least I will publish the PDF versions there.

I also need to add you to the relevant credits sections. Do you want to use your real name, or just be credited with the user name?
